### PR TITLE
Bump `vimeo/psalm` to 4.20

### PR DESCRIPTION
**Pull request by boesing**
Hey there,

this is the first patch of a series of patches I am planning to do for this package:

1. update `vimeo/psalm` to latest version
2. change usage of `$this->assert*` to `self::assert*` in unit tests
3. add support for PSR-6 caches

Would prefer having all patches mentioned above together in one of the next releases.

Thanks and let the review begin :D

### CODE REVIEW

File: test/ConnectionFactoryTest.php [hunk 1]
```diff
@@ -70,13 +82,9 @@ function (string $id) {
             foreach ($e->getTrace() as $entry) {
                 if (
                     isset($entry['class'])
-                    && in_array($entry['class'], [
-                        PDOMySQLDriver::class,
-                        AbstractMySQLDriver::class,
-                        ExceptionConverter::class,
```
Comment (Ocramius) [inline test/ConnectionFactoryTest.php:76]: Wait... `ExceptionConverter` is gone on purpose? :thinking:

Comment (boesing) [reply test/ConnectionFactoryTest.php:76]: Conditional, not sure if its only available in lower versions or higher versions. would wonder if it exists only in lower versions 🤷🏼‍♂️

File: test/ConnectionFactoryTest.php [hunk 2]
```diff
@@ -30,11 +31,22 @@ final class ConnectionFactoryTest extends TestCase
...
+        PDOMySQLDriver::class,
+        AbstractMySQLDriver::class,
+    ];
+
     public function setUp(): void
     {
         $this->configuration  = $this->createMock(Configuration::class);
         $this->eventManger    = $this->createMock(EventManager::class);
         $this->customPlatform = $this->createMock(AbstractPlatform::class);
+        if (! class_exists(ExceptionConverter::class)) {
```
Comment (boesing) [inline test/ConnectionFactoryTest.php]: todo: do not use return...

File: test/ConnectionFactoryTest.php (hunk 1 above)
Comment (Ocramius) [reply test/ConnectionFactoryTest.php:76]: @boesing could you raise the minimum version then (if not already the case)?

Comment (boesing) [reply test/ConnectionFactoryTest.php:76]: Sure, I could do that, but I would prefer not to do so due to the fact that it is not really necessary for the coverage of this project.

### TL;DR

- the `ExceptionConverter` was introduced in `doctrine/dbal` v3.0.0
- dropping support for `doctrine/dbal` v2 would also contain code cleanup and additions to this repo which I would prefer not to do in a `vimeo/psalm` update

---

So I could also just bump the `composer.lock` to contain `doctrine/dbal` `^3.0` and psalm won't complain anymore.
Easier change, less impact, same result.

WDYT?

Comment (Ocramius) [reply test/ConnectionFactoryTest.php:76]: Nope, either the code is valid for the ranges that we support, or it isn't :P 

Let's please bump dependencies.

Note: we bump often and eagerly in all of `roave/`. The reason there wasn't much activity so far is just because we haven't configured @dependabot for this repo (yet).

Comment (boesing) [reply test/ConnectionFactoryTest.php:76]: Okay, I'll stop working on this and focus on the PSR-6 support.
After this change, some deprecation quirks can be removed and its easier to drop support for `doctrine/dbal` <3

So I will remove the psalm changes from that upcoming PR to then rebase this PR off that PR.

--- 

But just to leave this here: 

this is kinda annoying and absolutely unnecessary from my current PoV. 
From a functionality aspect, this just takes more of my time than necessary. Without this PR, this won't be a problem and I doubt that it would change anytime soon - but now, as there are changes and you've seen it, you want me to put more effort in this. 

But I will do as you wish as this is your repository and therefore your rules. 

> Nope, either the code is valid for the ranges that we support, or it isn't :P

And it is. Thats why tests are passing on lowest and latest. Not sure what you want to say here but as I said: your repository, your rules.

Comment (Ocramius) [reply test/ConnectionFactoryTest.php:76]: Not sure what's problematic about raising minimum version according to this `ExceptionConverter`? :thinking:

**Comment by boesing** [comment]
Drafting this until #63 is merged so I can rebase this PR.

**Comment by Ocramius** [comment]
Can now be rebased

File: test/ConnectionFactoryTest.php (hunk 2 above)
Comment (Ocramius) [reply test/ConnectionFactoryTest.php]: Should this now be kept?

Comment (boesing) [reply test/ConnectionFactoryTest.php]: nope it can be dropped, good catch!

**Review by Ocramius — APPROVED** [review]
Patch got much smaller, thanks @boesing!
